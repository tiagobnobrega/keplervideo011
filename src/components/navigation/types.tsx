/*
 * Copyright (c) 2024 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 */

import { DrawerScreenProps } from '@amzn/react-navigation__drawer/lib/typescript/src/types';
import { StackNavigationProp } from '@amzn/react-navigation__stack';
import { StackScreenProps } from '@amzn/react-navigation__stack/lib/typescript/src/types';
import { ImageSourcePropType, TouchableOpacity } from 'react-native';
import { TitleData } from '../../types/TitleData';

export enum Screens {
  DEFAULT_SCREEN = 'Home',
  APP_DRAWER = 'AppDrawer',
  HOME_SCREEN = 'Home',
  DETAILS_SCREEN = 'Details',
  PLAYER_SCREEN = 'Player',
  SETTINGS_SCREEN = 'Settings',
  SEARCH_SCREEN = 'Search',
  SEARCH_RESULTS_SCREEN = 'SearchResultsScreen',
  FEEDBACK_SCREEN = 'FeedBackScreen',
}

export enum DrawerType {
  PERMANENT = 'permanent',
  FRONT = 'front',
  SLIDE = 'slide',
  BACK = 'back',
}

export type AppStackParamList = {
  [Screens.APP_DRAWER]: undefined;
  [Screens.HOME_SCREEN]: undefined;
  [Screens.SEARCH_SCREEN]: undefined;
  [Screens.SETTINGS_SCREEN]: undefined;
  [Screens.FEEDBACK_SCREEN]: undefined;
  [Screens.SEARCH_RESULTS_SCREEN]: {
    searchKeyword: string;
  };
  [Screens.DETAILS_SCREEN]: {
    data: TitleData;
    sendDataOnBack: (id: number | string) => void;
  };
  [Screens.PLAYER_SCREEN]: {
    data: TitleData;
    sendDataOnBack: () => void;
  };
};

export type AppStackScreenProps<T extends keyof AppStackParamList> =
  StackScreenProps<AppStackParamList, T>;

export type HomeScreenNavigationProps = StackNavigationProp<
  AppStackParamList,
  Screens.HOME_SCREEN
>;

export type AppDrawerParamList = {
  [Screens.HOME_SCREEN]: undefined;
  [Screens.SETTINGS_SCREEN]: undefined;
  [Screens.SEARCH_SCREEN]: undefined;
};

export type AppDrawerScreenProps<T extends keyof AppDrawerParamList> =
  DrawerScreenProps<AppDrawerParamList, T>;

declare global {
  namespace ReactNavigation {
    interface RootParamList extends AppStackParamList {}
  }
}

export interface ButtonConfig {
  onPress: () => void;
  image: ImageSourcePropType;
  label: string;
  ref?: React.RefObject<TouchableOpacity>;
  testID: string;
}
