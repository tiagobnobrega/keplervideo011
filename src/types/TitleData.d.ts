import { MediaItem } from '@amzn/keplerblocks-types';

export type MimeType =
  | 'text/plain'
  | 'text/vtt'
  | 'application/x-subtitle' // .srt
  | 'application/x-subtitle-vtt' // .vtt
  | 'application/x-subtitle-sami'
  | 'application/x-subtitle-tmplayer'
  | 'application/x-subtitle-mpl2'
  | 'application/x-subtitle-dks'
  | 'application/x-subtitle-qttext'
  | 'application/x-subtitle-lrc';

export type VideoFormat = 'MPD' | 'HLS' | 'MP4' | 'DASH';

interface DrmScheme {
  name: string;
  licenseUri: string;
  headerTag?: string;
  headerData?: string;
}
interface TextTrack {
  label: string;
  language: string;
  uri: string;
  mimeType: MimeType;
}

export interface TitleData extends MediaItem {
  videoUrl: string;
  categories: string[];
  channelID: string;
  posterUrl: string;
  format: string;
  drmScheme?: DrmScheme;
  textTrack?: TextTrack[];
  uhd: boolean;
  secure: boolean;
  aCodec?: string;
  vCodec?: string;
  rentAmount: string;
}
