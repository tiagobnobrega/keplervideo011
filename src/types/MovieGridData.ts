import { TitleData } from './TitleData';

export type MovieGridData = {
  heading: string;
  testID: string;
  data: () => TitleData[];
};
