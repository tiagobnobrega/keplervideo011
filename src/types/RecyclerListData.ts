import { StyleProp, ViewStyle } from 'react-native';
import { RecyclerListViewProps } from 'recyclerlistview';

export interface CustomRecyclerListViewProps extends RecyclerListViewProps {
  contentContainerStyle?: StyleProp<ViewStyle>;
  initialNumToRender: number | undefined;
}
