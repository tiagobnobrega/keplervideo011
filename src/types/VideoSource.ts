// adapted from GStreamer MIME Types (https://gstreamer.freedesktop.org/documentation/subparse/subparse.html?gi-language=c)
export type MimeType =
  | 'text/plain'
  | 'text/vtt'
  | 'application/x-subtitle' // .srt
  | 'application/x-subtitle-vtt' // .vtt
  | 'application/x-subtitle-sami'
  | 'application/x-subtitle-tmplayer'
  | 'application/x-subtitle-mpl2'
  | 'application/x-subtitle-dks'
  | 'application/x-subtitle-qttext'
  | 'application/x-subtitle-lrc';

export type VideoFormat = 'MPD' | 'HLS' | 'MP4' | 'DASH';

interface DrmScheme {
  name: string;
  licenseUri: string;
  headerTag?: string;
  headerData?: string;
}
interface TextTrack {
  label: string;
  language: string;
  uri: string;
  mimeType: MimeType;
}

export interface VideoSource {
  title: string;
  uri: string;
  format: VideoFormat;
  drmScheme?: DrmScheme;
  textTrack?: TextTrack[];
  uhd: boolean;
  secure: boolean;
  aCodec?: string;
  vCodec?: string;
}
