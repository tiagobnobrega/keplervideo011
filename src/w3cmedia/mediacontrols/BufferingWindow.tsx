/*
 * Copyright 2022-2023 Amazon.com, Inc. or its affiliates. All rights reserved.
 *
 * AMAZON PROPRIETARY/CONFIDENTIAL
 *
 * You may not use this file except in compliance with the terms and
 * conditions set forth in the accompanying LICENSE.TXT file.
 *
 * THESE MATERIALS ARE PROVIDED ON AN "AS IS" BASIS. AMAZON SPECIFICALLY
 * DISCLAIMS, WITH RESPECT TO THESE MATERIALS, ALL WARRANTIES, EXPRESS,
 * IMPLIED, OR STATUTORY, INCLUDING THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.
 */

import LottieView from '@amzn/lottie-react-native';
import React from 'react';
import { StyleSheet, View, ViewProps } from 'react-native';
import { COLORS } from '../../styles/Colors';
import { scaleUxToDp } from '../../utils/pixelUtils';

const BufferingWindow = React.memo((props: ViewProps) => {
  return (
    <View {...props} style={styles.container} testID="buffering-view">
      <LottieView
        testID="video-buffer-view"
        source={require('../../js/resources/loader.json')}
        autoPlay
        loop
        style={styles.loader}
      />
    </View>
  );
});

export const styles = StyleSheet.create({
  container: {
    width: '100%',
    height: '100%',
    top: 0,
    left: 0,
    position: 'absolute',
    flex: 1,
    backgroundColor: COLORS.BLACK,
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center',
    zIndex: 5,
  },
  loader: {
    width: scaleUxToDp(80),
    height: scaleUxToDp(320),
    alignSelf: 'center',
  },
});
export default BufferingWindow;
