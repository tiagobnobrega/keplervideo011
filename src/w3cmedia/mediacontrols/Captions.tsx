import { TVFocusGuideView } from '@amzn/react-native-kepler';
import React, { useCallback, useState } from 'react';
import { StyleSheet, View } from 'react-native';
import { CAPTION_DISABLE_ID, CAPTION_DISABLE_TEXT } from '../../constants';
import { COLORS } from '../../styles/Colors';
import { scaleUxToDp } from '../../utils/pixelUtils';
import CaptionMenuItem from './CaptionMenuItem';
import CaptionOptions from './CaptionOptions';
import { CaptionMenuProps } from './types/Captions';

export const CaptionMenu = ({
  captionMenuVisibility,
  video,
  setSelectedCaption,
}: CaptionMenuProps) => {
  const [captionID, setCaptionID] = useState<string>(CAPTION_DISABLE_ID);

  const removeCurrentCaption = useCallback(() => {
    if (captionID !== CAPTION_DISABLE_ID) {
      console.log('Disabling Captions');
      if (video && video.textTracks) {
        video.textTracks.getTrackById(captionID)!.mode = 'hidden';
      }
    }
    setCaptionID(CAPTION_DISABLE_ID);
  }, [captionID, video]);

  const enableCaption = useCallback(
    (id: string) => {
      if (id === captionID) {
        return;
      }
      removeCurrentCaption();
      if (video && video.textTracks) {
        video.textTracks.getTrackById(id)!.mode = 'showing';
        setCaptionID(id);
        setSelectedCaption(id);
      }
      setCaptionID(id);
      setSelectedCaption(id);
    },
    [captionID, removeCurrentCaption, video, setSelectedCaption],
  );

  const turnOffCaptions = useCallback(() => {
    setSelectedCaption(CAPTION_DISABLE_ID);
    removeCurrentCaption();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <TVFocusGuideView trapFocusLeft trapFocusRight trapFocusUp>
      {captionMenuVisibility ? (
        <View style={styles.menu}>
          <CaptionOptions
            video={video}
            enableCaption={enableCaption}
            selectedCaptionId={captionID}
          />
          <CaptionMenuItem
            key={CAPTION_DISABLE_TEXT}
            text={CAPTION_DISABLE_TEXT}
            onPress={turnOffCaptions}
            selected={captionID === CAPTION_DISABLE_ID}
            testID={'caption-menu-focus-element'}
          />
        </View>
      ) : null}
    </TVFocusGuideView>
  );
};

const styles = StyleSheet.create({
  menu: {
    backgroundColor: COLORS.BLACK + 'E6',
    width: scaleUxToDp(250),
    justifyContent: 'center',
    zIndex: 30,
    borderRadius: scaleUxToDp(15),
    overflow: 'hidden',
  },
  captionSelected: {
    backgroundColor: COLORS.DARK_GRAY + 'D9',
    borderRadius: scaleUxToDp(40),
  },
});
