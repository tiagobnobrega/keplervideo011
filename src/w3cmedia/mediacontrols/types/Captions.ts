import { VideoPlayer } from '@amzn/react-native-w3cmedia';

export interface CaptionButtonProps {
  onPress: () => void;
  video: VideoPlayer;
  captionMenuVisibility: boolean;
  testID: string;
  playerControlType?: string;
  isCaptionButtonFocused: boolean | null;
}

export interface CaptionOptionProps {
  video: VideoPlayer;
  enableCaption: (captionId: string) => void;
  selectedCaptionId: string | null;
}

export interface CaptionMenuItemProps {
  text: string;
  index?: string;
  selected?: boolean;
  onPress: () => void;
  testID?: string;
}

export interface CaptionMenuProps {
  captionMenuVisibility: boolean;
  video: VideoPlayer;
  setSelectedCaption: (id: string) => void;
}
