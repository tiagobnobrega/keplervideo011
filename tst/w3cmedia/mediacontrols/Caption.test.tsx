/*
 * Copyright (c) 2024 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 */
import { VideoPlayer } from '@amzn/react-native-w3cmedia';
import { describe } from '@jest/globals';
import '@testing-library/jest-native/extend-expect';
import { render } from '@testing-library/react-native';
import React from 'react';
import { CaptionMenu } from '../../../src/w3cmedia/mediacontrols/Captions';
import { CaptionMenuProps } from '../../../src/w3cmedia/mediacontrols/types/Captions';

const videoRef: React.MutableRefObject<VideoPlayer | null> = {
  current: {
    textTracks: {
      length: 1,
      getTrackById: jest.fn(id => ({
        mode: id === '0' ? 'showing' : 'hidden',
      })),
    },
  } as unknown as VideoPlayer,
};
const renderCaptionButton = (props?: Partial<CaptionMenuProps>) => {
  const setSelectedCaptionMock = jest.fn();

  return (
    <CaptionMenu
      captionMenuVisibility={true}
      setSelectedCaption={setSelectedCaptionMock}
      video={videoRef.current as VideoPlayer}
      {...props}
    />
  );
};
describe('ControlBar tests', () => {
  it('component renders correctly', () => {
    const component = render(renderCaptionButton());
    expect(component).toMatchSnapshot();
  });
  it('component renders correctly with captionMenuVisibility enabled', () => {
    const component = render(
      renderCaptionButton({ captionMenuVisibility: true }),
    );
    expect(component).toMatchSnapshot();
  });
  it('component renders correctly with captionMenuVisibility disabled', () => {
    const component = render(
      renderCaptionButton({ captionMenuVisibility: false }),
    );
    expect(component).toMatchSnapshot();
  });
  it('component renders correctly with textTracks 0 length', () => {
    const component = render(
      renderCaptionButton({
        video: {
          current: {
            ...videoRef.current,
            textTracks: {
              length: 0,
              getTrackById: jest.fn(),
            },
          } as unknown as VideoPlayer,
        } as unknown as VideoPlayer,
      }),
    );
    expect(component).toMatchSnapshot();
  });
});
describe('Functions gets call properly', () => {
  test('CaptionOptions useCallback', () => {
    render(renderCaptionButton({ captionMenuVisibility: true }));
    expect(videoRef.current?.textTracks.getTrackById('0')?.mode).toBe(
      'showing',
    );
    expect(videoRef.current?.textTracks.getTrackById('1')?.mode).toBe('hidden');
  });
});
