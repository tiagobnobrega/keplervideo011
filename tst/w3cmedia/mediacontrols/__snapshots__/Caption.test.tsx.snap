// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlBar tests component renders correctly 1`] = `null`;

exports[`ControlBar tests component renders correctly with captionMenuVisibility disabled 1`] = `null`;

exports[`ControlBar tests component renders correctly with captionMenuVisibility enabled 1`] = `null`;

exports[`ControlBar tests component renders correctly with textTracks 0 length 1`] = `null`;
